packages/test-sync/src/jvmTest/kotlin/io/realm/kotlin/jvm/RealmTests.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.kotlin.jvm
+
+import io.realm.kotlin.Realm
+import io.realm.kotlin.entities.sync.ChildPk
+import io.realm.kotlin.entities.sync.ParentPk
+import io.realm.kotlin.mongodb.Credentials
+import io.realm.kotlin.mongodb.sync.SyncConfiguration
+import io.realm.kotlin.test.mongodb.TestApp
+import io.realm.kotlin.test.util.TestHelper
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class RealmTests {
+
+    // Ensure that we cleanup all threads owned by Realm when closing the Realm as it will
+    // otherwise prevent things like the JVM main thread from exiting.
+    //
+    // Since both coroutines and OkHttp spin up a fair amount of workers with unpredictable
+    // lifecycle, it is very difficult to track all threads. Instead this test just makes a best
+    // effort in detecting the cases we do know about.
+    @Test
+    fun cleanupAllRealmThreadsOnClose() = runBlocking {
```

Comment:
Shouldn't we have a similar test on native?

Reply:
Ideally yes, but I don't think we have the same capabilities on Native to introspect the state of threads, so I assumed that for the things we are interested in, testing on JVM should be sufficient?

Reply:
Yeah, I guess this ensures that we are invoking the expected underlying framework calls, so this do ensure that our overall implementation should be right. But then we should maybe do a test that shows that we don't close custom dispatchers or create new threads if we pass them in?

Reply:
Good point. That is currently missing. I will add that 👍